### PR TITLE
Fixed bug in is_cannonial for polynomials

### PR DESCRIPTION
**Pull request by CodeMaxx**
Earlier `UnivariateIntPolynomial` allowed polynomials with symbol as `""`(empty) and with powers greater than 0.
```cpp
RCP<const UnivariateIntPolynomial> R = univariate_int_polynomial(none, {{5, 2_z}});
```
was treated as canonical and on printing as string gave `2*`

Error in`UnivariatePolynomial` was that when symbol was `""`(empty) and the `dict` was empty, they still checked for the first element of the `dict`(which did not exist).
So ```cpp
RCP<const UnivariateIntPolynomial> S = univariate_int_polynomial(none, {});
```
was not being considered canonical due to which this code below gave an error
```cpp
RCP<const UnivariatePolynomial> c = univariate_polynomial(none, 0, {{0, 5}});
REQUIRE(c->diff(x)->__str__() == "0");
```
This was due to the following code executed in `derivative.cpp` when symbol was `""`(empty)
```cpp
map_int_Expr d;
return make_rcp<const UnivariatePolynomial>(self.get_var(), 0, std::move(d)); 
// empty symbol, empty dict
```

@isuruf @irislq @chenchfort @Sumith1896

### CODE REVIEW

File: symengine/polynomial.cpp [hunk 1]
```diff
@@ -22,9 +22,8 @@ UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, c
 
 bool UnivariateIntPolynomial::is_canonical(const unsigned int &degree_, const map_uint_mpz& dict) const {
     if (var_->get_name() == "")
-        if (!(dict.empty() or dict.size() == 1))
+        if (!((dict.empty() or dict.size() == 1) and (dict.empty() or dict.begin()->first == 0)))
```
Comment (isuruf) [inline symengine/polynomial.cpp]: `dict.empty() or (dict.size() == 1 and dict.begin()->first == 0)` is simpler than above.

**Comment by CodeMaxx** [comment]
Ping @isuruf

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -347,8 +346,9 @@ UnivariatePolynomial::UnivariatePolynomial(const RCP<const Symbol> &var, const s
 
 bool UnivariatePolynomial::is_canonical(const unsigned int &degree_, const map_int_Expr& dict) const {
     if (var_->get_name() == "")
-        if (!((dict.empty() or dict.size() == 1) and dict.begin()->first == 0))
+        if (!((dict.empty() or dict.size() == 1) and (dict.empty() or dict.begin()->first == 0)))
```
Comment (isuruf) [inline symengine/polynomial.cpp]: Same comment as above

**Comment by irislq** [comment]
@isuruf @CodeMaxx CMake build issues

**Comment by CodeMaxx** [comment]
Travis is passing. See here https://travis-ci.org/CodeMaxx/symengine/builds/119992298 .<br/>
Some builds failed due to internal errors of Travis.

**Comment by irislq** [comment]
@CodeMaxx Ok, just wanted to make sure since it doesn't show up here

**Comment by isuruf** [comment]
Thanks for the PR

**Pull request merged**